File: Solutions/하루 한문제 코테준비/불량사용자_프로그래머스_64064.py
import copy
import re
from itertools import permutations
def check(origin,patterns):
    for idx,i in enumerate(patterns):
        p = re.compile(i)
        m = p.match(origin[idx])
        if(not m):
            return False
        if(m.group() != origin[idx]):
            return False
        # print(i,origin[idx],p.match(origin[idx]))
    return True



def solution(user_id, banned_id):
    answer = 0
    pattens = []
    for patten in banned_id:
        temp = ''
        for each in patten:
            if(each =='*'):
                temp += "[0-9a-z]"
            else:
                temp += each
        pattens.append(temp)
    # print(pattens)
    userList = permutations(user_id,len(pattens))
    # print(userList)
    result = []
    for i in userList:
        if(check(list(i),pattens)):
            if(set(i) not in result):
                result.append(set(i))
            # print(set(i),pattens)
    # print(result)
    return len(result)

print(solution(["frodo", "fradi", "crodo", "abc123", "frodoc"],["fr*d*", "abc1**"]))
print(solution(["frodo", "fradi", "crodo", "abc123", "frodoc"],["*rodo", "*rodo", "******"]))

File: Solutions/하루 한문제 코테준비/javascript_solve/불량사용자_프로그래머스_64064.js
function permutation(arr, selectNum) {
    let result = [];
    if (selectNum === 1) return arr.map((v) => [v]);
  
    arr.forEach((v, idx, arr) => {
      const fixer = v;
      const restArr = arr.filter((_, index) => index !== idx);
      const permuationArr = permutation(restArr, selectNum - 1);
      const combineFixer = permuationArr.map((v) => [fixer, ...v]);
      result.push(...combineFixer);
    });
    return result;
}
function check(origin,patterns){
    let flag = true;
    patterns.forEach((i,idx)=>{
        const p = new RegExp(i);
        const m = p.exec(origin[idx]);
        // console.log("rex",i,origin[idx],m)
        if(!m){
            // console.log("nulllllll")
            flag = false;
        }else{
            if(m[0] !== origin[idx]){
                flag = false;
            }
        }
        
        
    });
    return flag;
}
function solution(user_id, banned_id) {
    var answer = 0;
    const patterns = [];
    for(let i=0; i<banned_id.length; i++){
        let temp = '';
        const pattern = banned_id[i];
        for(let j=0; j<pattern.length; j++){
            const each = pattern[j];
            if(each==='*'){
                temp += '[0-9a-z]';
            }else{
                temp += each;
            }
        }
        patterns.push(temp);
    }
    const userList = permutation(user_id,patterns.length);
    const result = new Set();
    userList.forEach((val,idx)=>{
        const set = new Set(val)
        // console.log(val)
        if(check(Array.from(val),patterns)){
            const temp = val.sort().join('');
            if(!result.has(temp)){
                result.add(temp);
            }
        }
    });
    // console.log(result)
    return result.size;
}
console.log(solution(["frodo", "fradi", "crodo", "abc123", "frodoc"],["*rodo", "*rodo", "******"]));